src/main/scala/cognite/spark/v1/DefaultSource.scala

```diff
@@ -188,104 +188,108 @@ class DefaultSource
-        }
-
-      dataRepartitioned.foreachPartition((rows: Iterator[Row]) => {
-        import CdpConnector.ioRuntime
 
-        val maxParallelism = config.parallelismPerPartition
-        val batches = Stream.fromIterator[IO](rows, chunkSize = batchSize).chunks
+    val relation: CdfRelation = resourceType match {
+      case AssetHierarchyBuilder.name =>
+        new AssetHierarchyBuilder(config)(sqlContext)
+      case NumericDataPointsRelation.name =>
+        new NumericDataPointsRelationV1(config)(sqlContext)
+      case StringDataPointsRelation.name =>
+        new StringDataPointsRelationV1(config)(sqlContext)
+      case EventsRelation.name =>
+        new EventsRelation(config)(sqlContext)
+      case TimeSeriesRelation.name =>
+        new TimeSeriesRelation(config)(sqlContext)
+      case AssetsRelation.name =>
+        new AssetsRelation(config)(sqlContext)
+      case FilesRelation.name =>
+        new FilesRelation(config)(sqlContext)
+      case SequenceRelation.name =>
+        new SequencesRelation(config)(sqlContext)
+      case LabelsRelation.name =>
+        new LabelsRelation(config)(sqlContext)
+      case SequenceRowsRelation.name =>
+        createSequenceRows(parameters, config, sqlContext)
+      case RelationshipsRelation.name =>
+        new RelationshipsRelation(config)(sqlContext)
+      case DataSetsRelation.name =>
+        new DataSetsRelation(config)(sqlContext)
+      case FlexibleDataModelBaseRelation.name =>
+        createFlexibleDataModelRelation(parameters, config, sqlContext)
+      case _ => sys.error(s"Resource type $resourceType does not support save()")
+    }
 
-        val operation: Seq[Row] => IO[Unit] = config.onConflict match {
-          case OnConflictOption.Abort =>
-            relation.insert
-          case OnConflictOption.Upsert =>
-            relation.upsert
-          case OnConflictOption.Update =>
-            relation.update
+    relation match {
+      case relation: DataPointsRelationV1[_] =>
+        if (config.onConflict == OnConflictOption.Delete) {
+          // Datapoints support 100_000 per request when inserting, but only 10_000 when deleting
+          val batchSize = config.batchSize.getOrElse(Constants.DefaultDataPointsLimit)
+          data.foreachPartition((rows: Iterator[Row]) => {
+            import CdpConnector.ioRuntime
+            val batches = rows.grouped(batchSize).toVector
+            batches.parTraverse_(relation.delete).unsafeRunSync()
+          })
+        } else {
+          // datapoints need special handling of dataframes and batches
+          relation.insert(data, overwrite = true)
+        }
+        relation
+      case relation: AssetHierarchyBuilder =>
+        config.onConflict match {
           case OnConflictOption.Delete =>
-            relation.delete
+            relation.delete(data)
+          case _ =>
+            relation.buildFromDf(data)
         }
+        relation
+      case relation: CdfRelation with WritableRelation =>
```

Comment:
It looks to be on same level as cases by relation types so it will not be applied if any of previous ones match
May need to open in IDE old&new to double-check though

Reply:
Think you are misreading this? there is two pattern matching statements now.

First ones does the case by case for each source type and create the relation

Second one does some handling depending on the type of relation, there is special handling for both Datapoints and ... I forgot which one and all the rest go through this case here

Reply:
could be, in web UI it is a bit confusing to see the nesting, will try in a proper UI

Reply:
Definitely :)

Reply:
Ah you mean it won't be applied for Datapoints and AssetHierarchyBuilder? That is true, but it's the same as before, as it was before nested in the "else" statement

Reply:
Maybe it should be applied in these cases too, but that would be a behavior changes and should be looked at separately.

Reply:
yep, for this PR best to keep existing behavior